Patch Environment On OSX And Allow A Different Environment Per Window

lee-dohm opened the pull request:
- [X] ~~Create an `atom.project.getEnv` API~~
- [X] Fixup `PATH` on OS X when `TERM` is not set
- [X] Patch `process.env` on startup
- [x] Receive passed in environment variables on subsequent launches

Fixes #11037
Fixes https://github.com/atom/atom/issues/6956

lee-dohm committed:
Add basic atom.project.getEnv API

lee-dohm committed:
:memo: Add documentation for atom.project.getEnv

lee-dohm committed:
Move new functions to private section

thomasjo commented:
![giphy](https://cloud.githubusercontent.com/assets/3622/13520452/af2508d8-e1e0-11e5-95f1-bcda5644a66f.gif)

steelbrain commented on src/project.coffee:
```diff
@@ -271,9 +274,51 @@ class Project extends Model
+
+  getShellEnv: ->
+    shell = process.env.SHELL ? "/bin/bash"
+    results = child_process.spawnSync shell, ["--login"], input: "env", encoding: "utf8"
```
I just tried it and it doesn't give accurate terminal results for `$PATH` me on my mac
```
# From terminal
/usr/local/bin:/usr/local/sbin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/steel/n/bin:/usr/games/bin
```
```
# From executing this
/usr/local/bin:/usr/local/sbin:/usr/bin:/bin:/usr/sbin:/sbin
```

steelbrain replied:
We ran into the same problem when working with linters, we ended up creating [`consistent-path`](https://www.npmjs.com/package/consistent-path) and [`consistent-env`](https://www.npmjs.com/package/consistent-env). Both are very battle tested and used by almost all of the linter providers.

`consistent-path` is a `consistent-env` dependent package which provides just the PATH, but looks like you could use `consistent-env` here

thomasjo replied:
@lee-dohm If you replace the `--login` flag with `--interactive` this should work as intended :sparkles:

steelbrain replied:
@thomasjo That would break linux because Atom is spawned on linux with `nohup`, see https://github.com/steelbrain/nohup-shell-hanging-bug

joefitzgerald replied:
@steelbrain the shell launching should not be done on any platform other than OS X. Consequently, nothing should break on Linux.

steelbrain replied:
@joefitzgerald There are several pairs of Desktop Environments and shells on Linux that do not supply ENV correctly to apps not run from terminal, most notably GNOME Shell and ZSH. Therefore we do need to fix it there as well

joefitzgerald replied:
Well we don't need to in this PR; this PR is intended to solve for the OS X issue, and add a project level API for the environment.

That doesn't preclude more cases beyond OS X; can you please open up (or link) an issue for each additional scenario, describing steps to reproduce, and desired behavior?

thomasjo replied:
@steelbrain Correct me if I'm wrong, but that has nothing to do with the DE nor the shell. What you're describing would be distribution-specific configuration problems. There are lots of silly *nix distributions out there with horrible defaults.

But if this truly is a problem (in particular the HUP scenario described earlier), then we'll just have to deal with it.

thomasjo replied:
> Well we don't need to in this PR; this PR is intended to solve for the OS X issue, and add a project level API for the environment.

Agreed! We can patch other problems at a later point, once we can reliably reproduce the anecdotal reports.

steelbrain replied:
@joefitzgerald @thomasjo What do you guys have against the npm modules I suggested? It takes care of all the scenarios and fixes this for everyone not just the handful few

thomasjo commented on spec/fixtures/sample.txt:
```diff
@@ -1 +1 @@
-Some textSome textSome text.
+Some textSome textSome textSome text.
```
This change seems unintentional. Revert?

lee-dohm replied:
Thanks, yeah ... not sure why that happens on my machine :angry:

mnquintana replied:
Do we have a spec that inadvertently edits that fixture? :eyes:

lee-dohm replied:
I suspect we do, yes.

lee-dohm committed:
:fire: Revert change to test fixture

lee-dohm committed:
Add --interactive option to get the user's full PATH with all the trimmings

lee-dohm committed:
:memo: Add Finder to the list of ways an app can be launched

lee-dohm committed:
:white_check_mark: Add test for error handling

lee-dohm committed:
Patch process.env on startup

lee-dohm commented:
This is still not complete but I wanted to push up what I have so far. I decided to include patching `process.env` on startup per @joshaber's suggestion on #11037 (though the approach I chose is probably wrong). I also did a bit more reading on #4126 (and there's still #6956 that I haven't caught up on in a while). There are a lot of nuances that will need to be covered for this to be acceptable. I'm going to keep plugging away at it :grinning:

joefitzgerald commented on src/atom-environment.coffee:
```diff
@@ -229,6 +229,10 @@ class AtomEnvironment extends Model
 
+    # Patch the `process.env` on startup to fix the problem first documented
+    # in #4126
+    process.env = @project.getEnv()
```
This is going to make it less clean, but this should happen prior to packages being loaded. I.e. prior to @project being defined. Think of all the activities that packages may do on activation that they _should not be doing_ (but still are doing)n that are dependent on `process.env`.

joefitzgerald replied:
If you move the patching logic such that it is executed in the first atom-environment, then maybe `@project.getEnv()` could be renamed to be `@project.env`. Perhaps we could also externalize the logic into a file that is only required if patching is needed (`javascript` FTW?).

lee-dohm replied:
Thanks for the ideas. I'll take a deeper look at this later tonight or tomorrow.

lee-dohm replied:
> then maybe `@project.getEnv()` could be renamed to be `@project.env`

That would lose us the current feature that packages can't muck up the environment for other packages (at least through the public API). Right now, everyone gets a copy and then they can do what they want with their copy.

I'll definitely move the first initialization further up though.

joefitzgerald replied:
Yeah, fair enough on making it a function:

* `project.getEnv()`?
* `project.env()`?
* `project.environment()`?

lee-dohm replied:
Looking deeper, it appears that packages aren't loaded until `AtomEnvironment.startEditorWindow`:

https://github.com/atom/atom/blob/ld-osx-environment/src/atom-environment.coffee#L674

So doing the patching of the environment in the `AtomEnvironment` constructor does make it happen before packages are loaded.

lee-dohm replied:
Personally, I was sticking with just "env" as the main part of the name to make it familiar to Node developers and `process.env`. I added "get" to the front to make it clear it is a function rather than just an attribute (and all of the other public APIs have a verb in the front). I can totally be swayed though.

joefitzgerald commented on src/project.coffee:
```diff
@@ -271,9 +274,66 @@ class Project extends Model
+    unless @env?
+      @env = _.clone(process.env)
+      if process.platform is "darwin" and not process.env.TERM?
+        shellPath = @getShellPath()
```
We shouldn't just be setting the path; we should swap out the entire environment.

thomasjo replied:
> We shouldn't just be setting the path; we should swap out the entire environment.

I agree fully with this. The goal here should be to make sure all launch experience are equivalent.

lee-dohm replied:
Done.

lee-dohm committed:
Add environment module for getting environment info

lee-dohm committed:
Update Project to use new environment module

lee-dohm committed:
Fix broken command parameter

joefitzgerald commented on src/environment.coffee:
```diff
@@ -0,0 +1,41 @@
+child_process = require('child_process')
```
Why are we creatomg new files in `.coffee`? Want me to convert to `.js`?

thedaniel replied:
yeah this should probably be js

nathansobo commented:
I don't think we should make a new API for fetching environment variables. I think we should extend the existing API so it works correctly despite our multi-process setup, which shouldn't be that relevant to Atom users. The only justifications I could see for inventing something new would be the following:

1) Someone might write into `process.env`. :shrug: People can always write places they shouldn't. Doesn't justify building a new construct in my mind.
2) People don't expect `process.env` to have these particular variables. I think we can reinterpret a bit since we own the platform.

joefitzgerald commented:
Yep @nathansobo i've already made the requisite changes :) Pushing the commits shortly.

joefitzgerald committed:
Remove Project API, Work With process.env Directly

- Convert environment.coffee and environment-spec.coffee to JavaScript
- Pass the process's environment across the wire when launching atom multiple times from the command line

joefitzgerald committed:
Remove Cruft

joefitzgerald committed:
Unfocus Spec

nathansobo commented on spec/environment-spec.js:
```diff
@@ -0,0 +1,161 @@
+    delete process._originalEnv
+    options = {
+      platform: process.platform,
+      env: _.clone(process.env)
```
`Object.assign`?

lee-dohm replied:
I'll try it out. I haven't used it before. If it works the same, I'll update it.

nathansobo commented on src/atom-environment.coffee:
```diff
@@ -4,6 +4,7 @@ path = require 'path'
 
 _ = require 'underscore-plus'
 {deprecate} = require 'grim'
+environment = require('./environment')
```
This is a weird name to me... what do you think of `environmentHelpers` or something?

lee-dohm replied:
Works for me, I'll change it.

lee-dohm committed:
Rename environment -> environmentHelpers

lee-dohm committed:
Swap _.clone for Object.assign

joefitzgerald commented:
:tada: :joy_cat:

basnijholt commented:
Finally 👍😄

sbromberger commented:
Apparently this didn't make 1.6.0. Will we see it in 1.6.1? (Fingers crossed....)

joefitzgerald commented:
It's in the beta channel. You can run stable and beta side by side. atom.io/beta

sbromberger commented:
Ah, thanks. I hesitate to run beta since Atom is now a huge part of my workflow and I don't really want to disrupt / switch between things. If I understand the blog post, this feature will be merged into -stable at some point. Since it's in -beta now, does that mean 1.6.1?

thomasjo commented:
> Since it's in -beta now, does that mean 1.6.1

@sbromberger 1.7.0 (unless it's reverted for whatever reason).

lierdakil commented:
> Should -i be used when spawning the shell?

I found that many Linux/OSX users tend to customize their environment in rc file (`.bashrc`/`.zshrc`), which only works for *interactive* shells. This is certainly not a preferred way, but it's used by many people nonetheless. Not all of them understand differences between login and interactive shell, and not all of them care.

It's true that running interactive shell can lead to problems. But running login shell can too. For example, I run `ssh-agent` from `.profile` -- not something you want to do when starting Atom (I'm on Linux though, so make of it what you will)

ioquatix commented:
Hey, since I tried to solve this problem in [script-runner](https://atom.io/packages/script-runner) I thought I'd chime in here.

Firstly, for people running things like RVM, or the equivalent for python, you definitely want to get the environment from an interactive login shell, or you won't get the right things.

We tried tied to extract this into a separate module, called [shell-environment](https://www.npmjs.com/package/shell-environment).

We also had some issues on Linux where things like DISPLAY were not getting set and this might be unintuitive for users who are expecting things to "just work".

If you look at the approach we took in shell-environment, you'll see we avoided many of the issues of using an interactive shell by piping the results of `env` back through a custom pipe.

gnestor commented:
I have discovered another minutia missing from this patch. Unless Atom is launched from the terminal, `process.env.PWD` is set to '/'. If `env` (bash) is run as a child/buffered process in Atom, PWD is correctly set to my project's root directory. Therefore, it appears like the method used in this PR to get the shell environment is failing for PWD. It's worth mentioning that PWD can be inferred from `atom.workspace.project.getPaths()[0]`. @lee-dohm What do you say?

lee-dohm commented:
@gnestor I think `PWD` is set correctly. It is set by the environment that Atom is launched within. The goal of this fix isn't to make the environment identical to what it would be if it were launched from the terminal at some pre-determined location. It is to import the user's standard profile. This change achieved that.

gnestor commented:
@lee-dohm Ok so it sounds like this is a separate issue. Do you agree though that Atom should correctly set `process.env.PWD` to the project's root (e.g. `atom.workspace.project.getPaths()[0]`) vs. "/" when Atom is launched outside of the terminal? If so, I will open a new issue.

lee-dohm commented:
@gnestor No, I don't agree that `PWD` should be set to the project's root. The concept of this change wasn't to make it as if someone launched from the project directory but to import the user's standard environment (as much as possible).

driskell commented:
In most of 1.12 betas (I'm on 1.12.0-beta7 at the moment on Sierra 10.12) there's a problem with new windows.

If you open from OS X dock and inspect process.env, all is fine and wonderful in the world/window. When you go to File -> New Window, and inspect process.env on the new window, it does not include any of the environment variables, anything from login is gone. I haven't had change to inspect things but seems might be an issue with inheriting environment when creating new windows. I can't find a related issue to this either.

thomasjo commented:
I can reproduce the behavior described by @driskell. Unsure whether this is a regression or not — was this specific workflow tested/fixed? /cc @lee-dohm @atom/feedback

driskell commented:
It looks like the new-window events etc don't pass in the current `env` to the `openPaths` (https://github.com/atom/atom/blob/e5cfc6b6e4b36b6c443c0526bfb4c816c666b0f2/src/main-process/atom-application.coffee#L210) so it doesn't do any environment update, unlike with the first window opening (or window opened from command line), which does. It also means dragging a folder onto dock to open a project means you get no env.

I'll raise a new issue for it this week, maybe even attempt a fix.

lee-dohm commented:
@driskell Have you filed an Issue on this behavior?

flujan commented:
Same issue here.

joefitzgerald commented:
@flujan have you updated to the latest version of atom? A fix for the issue described by @driskell was included in the latest patch.

flujan commented:
Yes, when I start ATOM from the terminal, it works. When I try to launch from the Dock it does not have the Path configured.

joefitzgerald commented:
@flujan https://github.com/atom/atom/pull/13200 your package(s) may need to be updated to make use of the `core:loaded-shell-environment` activationHook. Which package(s) are complaining abou the path?

flujan commented:
Guys, the Iterm2 plugin was causing the bug. Thanks.
